### ISSUE TITLE
about adding runGroupBy* functionality

### ISSUE THREAD

**Issue by wookietreiber** [functional-streams-for-scala/fs2#584]
## tl;dr

The implementation I currently work with looks as follows:

```scala
import scalaz.Monoid
import scalaz.concurrent.Task
import scalaz.std.list._
import scalaz.std.map._
import scalaz.stream._

implicit class RichProcess[O](underlying: Process[Task,O]) {

  def runGroupByFoldMap[K,O2](f: O => K)(g: O => O2)(implicit MO: Monoid[O2]): Task[Map[K,O2]] =
    underlying runFoldMap { cur =>
      Map(f(cur) -> g(cur))
    }

  def runGroupByFoldMonoid[K](f: O => K)(implicit MO: Monoid[O]): Task[Map[K,O]] =
    runGroupByFoldMap(f)(identity)

  def runGroupBy[K](f: O => K): Task[Map[K,List[O]]] =
    runGroupByFoldMap(f)(List(_))

}
```

If you think this is worth including in fs2, please let me know. I will then prepare a pull request against `topic/redesign`.

If you can think of another way to solve this, I'll be glad to hear about it.

## implementation notes

-   based on scalaz-stream 0.8.1

-   I have fixed a lot of the types because a) to make the implementation easier and b) I need only those types for my use case (see below)

-   all variants have the potential to cause out-of-memory errors for large streams (similar to `runLog`)

    this depends on the monoid, whether it really reduces the number of values, e.g. sum up numbers, or rather concatenates them, e.g. list concatenation

    I included `runGroupBy` for completeness sake - in practice, I guess, I wouldn't use it because of the inherent memory issues. Instead, I usually have some kind of ADT I use to count stuff and, naturally, there is a monoid for this ADT. So, my result is `Map[K,ADT]` and with a reasonable amount of keys there are no memory issues.

-   the long names may be a bit weird but I tried to stay as close with the naming to already existing functions

-   everything is implemented in terms of `runGroupByFoldMap` as it is the most general case

## my current use case

I'm using [strace](https://en.wikipedia.org/wiki/Strace) to analyze I/O behaviour of Linux processes. strace generates (potentially large) log files, one line per syscall. I am particularly interested in the Linux I/O syscalls `open`, `read`, `write` and `close`. I need `open` and `close` for (numeric) file descriptor to file name/path associations, and `read` and `write` to see the amount of bytes that are processed in a specific amount of time.

One of my analysis commands shows how much read/write operations with how many bytes in how many seconds have been performed **per file**, e.g.:

```
write 1G in 6.162 ms (~ 162G / s) with 1024 ops (~ 1M / op, ~ 1M request size) /dev/null
```

To generate the above output I need to do the following:

1.  read the file line by line
2.  parse the syscalls, collect relevant ones
3.  group by file
4.  sum up amount of bytes, seconds, etc. per read (resp. write) for each file
5.  print the per file summary

Steps 3. and 4. can be done with `runGroupByFoldMap`:

```scala
val analysis: Map[File,ADT] = parsed.runGroupByFoldMap(_.file)(ADT(_)).run
```

---

The project is called [strace-analyzer](https://github.com/wookietreiber/strace-analyzer) and I am currently working on converting it from using `scala.io.Source` to scalaz-stream/fs2 on the [wip-streams branch](https://github.com/wookietreiber/strace-analyzer/tree/wip-streams). Notable source locations in the latest commit are:

-   [runGroupBy* implementation](https://github.com/wookietreiber/strace-analyzer/blob/4db9860e0d364fae9a22ff23d5776355c1b3bc12/src/main/scala/util.scala#L35)
-   [read logs line by line (initial stream creation)](https://github.com/wookietreiber/strace-analyzer/blob/4db9860e0d364fae9a22ff23d5776355c1b3bc12/src/main/scala/analysis.scala#L38)
-   [parsing syscall entries to internal representation](https://github.com/wookietreiber/strace-analyzer/blob/4db9860e0d364fae9a22ff23d5776355c1b3bc12/src/main/scala/analysis.scala#L52)
-   [ADT](https://github.com/wookietreiber/strace-analyzer/blob/4db9860e0d364fae9a22ff23d5776355c1b3bc12/src/main/scala/file-op-summary.scala#L29)
-   [final per file I/O analysis](https://github.com/wookietreiber/strace-analyzer/blob/4db9860e0d364fae9a22ff23d5776355c1b3bc12/src/main/scala/io.scala#L30)

---

**Edit:**

-   added the `run` prefix to the `groupBy*` methods above

-   I have given `groupBy` a bit more thought, its signature should actually lock like:

    ```scala
      def groupBy[K](f: O => K): Process[Task, Map[K, Process[Task,List[O]]]] =
        ???
    ```

    I don't know how to implement this yet or if it is even possible to dynamically create streams of streams when new keys are discovered.

**Comment by refried** [comment]
Possibly dumb question: Does `runGroupByFoldMap` actually use `Monoid[O2].zero` at some level, or can it take a semigroup?

**Comment by wookietreiber** [comment]
@refried The `runGroupBy*` functions need a Monoid because `runFoldMap`, which is used underneath, needs a Monoid.

**Comment by fiadliel** [comment]
There is no `Monoid` in FS2 (0.9). You might consider https://github.com/functional-streams-for-scala/fs2-cats ? But probably some of the more basic Monoid folds should be implemented first.

**Comment by wookietreiber** [comment]
I see. There is no `runFoldMap` in fs2-core anymore. I would essentially need to add `runFoldMap` to fs2-cats as part of an implict class that wraps the new fs2 `Stream`, correct?

**Comment by refried** [comment]
Maybe this is a good opportunity to only require Semigroup, if that is possible :)

**Comment by wookietreiber** [comment]
FYI: I have opened a related issue on fs2-cats: functional-streams-for-scala/fs2-cats#1.

**Comment by wookietreiber** [comment]
@refried There was `foldSemigroup` in scalaz-stream, which is now also in PR functional-streams-for-scala/fs2-cats#2 (implemented in terms of `Stream.reduce`). However, there is no `Stream.runReduce` or similar, so there is (apparently) no way of implementing the `run*` counterpart without providing an initial zero/empty element for it to be applicable to `Semigroup` (at least as far as I can tell).

**Comment by wookietreiber** [comment]
Added in functional-streams-for-scala/fs2-cats#4.
